server/internal/core/application/indexer.go

```diff
@@ -0,0 +1,553 @@
+package application
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"fmt"
+	"github.com/ark-network/ark/common"
+	"github.com/ark-network/ark/common/tree"
+	"github.com/ark-network/ark/server/internal/core/domain"
+	"github.com/ark-network/ark/server/internal/core/ports"
+	"github.com/btcsuite/btcd/btcec/v2/schnorr"
+	"github.com/btcsuite/btcd/btcutil/psbt"
+	"github.com/decred/dcrd/dcrec/secp256k1/v4"
+	"math"
+	"sort"
+	"strings"
+	"time"
+)
+
+const (
+	maxPageSizeVtxoTree       = 300
+	maxPageSizeConnector      = 300
+	maxPageSizeForfeitTxs     = 500
+	maxPageSizeSpendableVtxos = 100
+	maxPageSizeTxHistory      = 200
+	maxPageSizeVtxoChain      = 100
+	maxPageSizeVirtualTxs     = 100
+)
+
+type IndexerService interface {
+	GetCommitmentTxInfo(ctx context.Context, txid string) (*CommitmentTxResp, error)
+	GetVtxoTree(ctx context.Context, req VtxoTreeReq) (*VtxoTreeResp, error)
+	GetForfeitTxs(ctx context.Context, req ForfeitTxsReq) (*ForfeitTxsResp, error)
+	GetConnectors(ctx context.Context, req ConnectorsReq) (*ConnectorResp, error)
+	GetSpendableVtxos(ctx context.Context, req SpendableVtxosReq) (*SpendableVtxosResp, error)
+	GetTransactionHistory(ctx context.Context, req TxHistoryReq) (*TxHistoryResp, error)
+	GetVtxoChain(ctx context.Context, req VtxoChainReq) (*VtxoChainResp, error)
+	GetVirtualTxs(ctx context.Context, req VirtualTxsReq) (*VirtualTxsResp, error)
+	GetSweptCommitmentTx(ctx context.Context, txid string) (*SweptCommitmentTxResp, error)
+}
+
+type indexerService struct {
+	pubkey *secp256k1.PublicKey
+
+	repoManager ports.RepoManager
+}
+
+func NewIndexerService(
+	pubkey *secp256k1.PublicKey,
+	repoManager ports.RepoManager,
+) IndexerService {
+	return &indexerService{
+		pubkey:      pubkey,
+		repoManager: repoManager,
+	}
+}
+
+func (i *indexerService) GetCommitmentTxInfo(
+	ctx context.Context, txid string,
+) (*CommitmentTxResp, error) {
+	round, err := i.repoManager.Rounds().GetRoundWithTxid(ctx, txid)
```

Comment:
This API can take time if the database and the round are big, we should use an ad hoc api that basically returns the data as defined in the proto.

Reply:
Done